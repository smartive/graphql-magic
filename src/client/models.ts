export type ClientEntity = Record<string, unknown>;
