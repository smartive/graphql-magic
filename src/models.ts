import { DateTime } from 'luxon';
import { Field } from '.';
import type { Context } from './context';
import type { OrderBy } from './resolvers/arguments';
import type { Value } from './values';

export type RawModels = RawModel[];

export type RawModel = ScalarModel | EnumModel | RawEnumModel | InterfaceModel | ObjectModel | RawObjectModel;

type BaseModel = {
  name: string;
  plural?: string;
  description?: string;
};

export type ScalarModel = BaseModel & { type: 'scalar' };

export type EnumModel = BaseModel & { type: 'enum'; values: string[]; deleted?: true };

export type RawEnumModel = BaseModel & { type: 'raw-enum'; values: string[] };

export type InterfaceModel = BaseModel & { type: 'interface'; fields: ModelField[] };

export type RawObjectModel = BaseModel & {
  type: 'raw';
  fields: RawObjectField[];
};

export type Entity = Record<string, unknown> & { createdAt?: DateTime; deletedAt?: DateTime };

export type Action = 'create' | 'update' | 'delete' | 'restore';

export type MutationHook = (
  model: Model,
  action: Action,
  when: 'before' | 'after',
  data: { prev: Entity; input: Entity; normalizedInput: Entity; next: Entity },
  ctx: Context
) => Promise<void>;

export type ObjectModel = BaseModel & {
  type: 'object';
  interfaces?: string[];
  queriable?: boolean;
  listQueriable?: boolean;
  creatable?: boolean | { createdBy?: Partial<RelationField>; createdAt?: Partial<DateTimeField> };
  updatable?: boolean | { updatedBy?: Partial<RelationField>; updatedAt?: Partial<DateTimeField> };
  deletable?:
    | boolean
    | { deleted?: Partial<BooleanField>; deletedBy?: Partial<RelationField>; deletedAt?: Partial<DateTimeField> };
  displayField?: string;
  defaultOrderBy?: OrderBy;
  fields: ModelField[];

  // temporary fields for the generation of migrations
  deleted?: true;
  oldName?: string;
};

type BaseNumberType = {
  unit?: 'million';
  min?: number;
  max?: number;
};

type BaseField = Omit<Field, 'type'>;

type PrimitiveField =
  | { type: 'ID' }
  | { type: 'Boolean' }
  | {
      type: 'String';
      stringType?: 'email' | 'url' | 'phone';
      large?: true;
      maxLength?: number;
    }
  | {
      type: 'DateTime';
      dateTimeType?: 'year' | 'date' | 'datetime' | 'year_and_month';
      endOfDay?: boolean;
    }
  | ({
      type: 'Int';
      intType?: 'currency';
    } & BaseNumberType)
  | ({
      type: 'Float';
      floatType?: 'currency' | 'percentage';
      double?: boolean;
      precision?: number;
      scale?: number;
    } & BaseNumberType)
  | { type: 'Upload' };

type RawObjectField = BaseField & PrimitiveField;

export type ModelField = BaseField &
  (
    | PrimitiveField
    | { type: 'json'; typeName: string }
    | { type: 'enum'; typeName: string; possibleValues?: Value[] }
    | { type: 'raw'; typeName: string }
    | {
        type: 'relation';
        typeName: string;
        toOne?: boolean;
        reverse?: string;
        foreignKey?: string;
        onDelete?: 'cascade' | 'set-null';
      }
  ) & {
    primary?: boolean;
    unique?: boolean;
    filterable?:
      | boolean
      | {
          default?: Value;
        };
    searchable?: boolean;
    orderable?: boolean;
    comparable?: boolean;
    queriable?:
      | boolean
      | {
          roles?: string[];
        };
    creatable?:
      | boolean
      | {
          roles?: string[];
        };
    updatable?:
      | boolean
      | {
          roles?: string[];
        };
    generated?: boolean;
    // The tooltip is "hidden" behind an icon in the admin forms
    tooltip?: string;
    defaultValue?: string | number | ReadonlyArray<string> | undefined;
    // If true the field must be filled within forms but can be null in the database
    required?: boolean;
    indent?: boolean;
    // If true the field is hidden in the admin interface
    hidden?: boolean;

    // temporary fields for the generation of migrations
    deleted?: true;
    oldName?: string;
  };

export type IDField = Extract<ModelField, { type: 'ID' }>;
export type BooleanField = Extract<ModelField, { type: 'Boolean' }>;
export type StringField = Extract<ModelField, { type: 'String' }>;
export type DateTimeField = Extract<ModelField, { type: 'DateTime' }>;
export type IntField = Extract<ModelField, { type: 'Int' }>;
export type FloatField = Extract<ModelField, { type: 'Float' }>;
export type JsonField = Extract<ModelField, { type: 'json' }>;
export type EnumField = Extract<ModelField, { type: 'enum' }>;
export type RawField = Extract<ModelField, { type: 'raw' }>;
export type RelationField = Extract<ModelField, { type: 'relation' }>;

export type Models = Model[];

export type Model = ObjectModel & {
  fieldsByName: Record<string, ModelField>;
  relations: Relation[];
  relationsByName: Record<string, Relation>;
  reverseRelations: ReverseRelation[];
  reverseRelationsByName: Record<string, ReverseRelation>;
};

export type Relation = {
  field: RelationField;
  model: Model;
  reverseRelation: ReverseRelation;
};

export type ReverseRelation = RelationField & {
  model: Model;
  field: RelationField;
  fieldModel: Model;
};
