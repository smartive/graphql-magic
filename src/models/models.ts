import { pluralize, singularize } from 'inflection';
import cloneDeep from 'lodash/cloneDeep';
import kebabCase from 'lodash/kebabCase';
import omit from 'lodash/omit';
import pick from 'lodash/pick';
import { isRelation } from '.';
import {
  CustomFieldDefinition,
  EnumFieldDefinition,
  FloatFieldDefinition,
  IDFieldDefinition,
  InputModelDefinition,
  IntFieldDefinition,
  InterfaceModelDefinition,
  JsonFieldDefinition,
  ObjectModelDefinition,
  OrderBy,
  PrimitiveFieldDefinition,
  RawEnumModelDefinition,
  StringFieldDefinition,
  UploadFieldDefinition,
} from '..';
import {
  BooleanFieldDefinition,
  DateTimeFieldDefinition,
  EntityFieldDefinition,
  EntityModelDefinition,
  EnumModelDefinition,
  ModelDefinition,
  ModelDefinitions,
  ObjectFieldDefinition,
  RelationFieldDefinition,
} from './model-definitions';
import { get, getLabel, summonByName, typeToField } from './utils';

// These might one day become classes

export type ObjectField = ObjectFieldDefinition;
export type EntityField = EntityFieldDefinition;
export type PrimitiveField = PrimitiveFieldDefinition;
export type IDField = IDFieldDefinition;
export type BooleanField = BooleanFieldDefinition;
export type StringField = StringFieldDefinition;
export type DateTimeField = DateTimeFieldDefinition;
export type IntField = IntFieldDefinition;
export type FloatField = FloatFieldDefinition;
export type UploadField = UploadFieldDefinition;
export type JsonField = JsonFieldDefinition;
export type EnumField = EnumFieldDefinition;
export type CustomField = CustomFieldDefinition;
export type RelationField = Omit<RelationFieldDefinition, 'foreignKey'> & { foreignKey: string };

export const getFullDefinitions = (modelDefinitions: ModelDefinitions) => {
  const definitions = cloneDeep(modelDefinitions);
  definitions.push(
    {
      kind: 'scalar',
      name: 'DateTime',
    },
    { kind: 'scalar', name: 'Upload' },
    {
      kind: 'raw-enum',
      name: 'Order',
      values: ['ASC', 'DESC'],
    },
  );
  const entities = definitions.filter(isEntityModelDefinition);

  for (const entity of entities) {
    if (entity.root) {
      definitions.push({
        kind: 'enum',
        name: `${entity.name}Type`,
        values: entities.filter((subModel) => subModel.parent === entity.name).map((subModel) => subModel.name),
      });
    }

    entity.fields = [
      {
        name: 'id',
        type: 'ID',
        nonNull: true,
        unique: true,
        primary: true,
        generated: true,
      },
      ...(entity.root
        ? [
            {
              name: 'type',
              kind: 'enum',
              type: `${entity.name}Type`,
              nonNull: true,
              generated: true,
            } satisfies EntityFieldDefinition,
          ]
        : []),
      ...entity.fields,
      ...(entity.creatable
        ? [
            {
              name: 'createdAt',
              type: 'DateTime',
              nonNull: true,
              orderable: true,
              generated: true,
              ...(typeof entity.creatable === 'object' && entity.creatable.createdAt),
            } satisfies DateTimeFieldDefinition,
            {
              name: 'createdBy',
              kind: 'relation',
              type: 'User',
              nonNull: true,
              reverse: `created${getModelPlural(entity)}`,
              generated: true,
              ...(typeof entity.creatable === 'object' && entity.creatable.createdBy),
            } satisfies RelationFieldDefinition,
          ]
        : []),
      ...(entity.updatable
        ? [
            {
              name: 'updatedAt',
              type: 'DateTime',
              nonNull: true,
              orderable: true,
              generated: true,
              ...(typeof entity.updatable === 'object' && entity.updatable.updatedAt),
            } satisfies DateTimeFieldDefinition,
            {
              name: 'updatedBy',
              kind: 'relation',
              type: 'User',
              nonNull: true,
              reverse: `updated${getModelPlural(entity)}`,
              generated: true,
              ...(typeof entity.updatable === 'object' && entity.updatable.updatedBy),
            } satisfies RelationFieldDefinition,
          ]
        : []),
      ...(entity.deletable
        ? [
            {
              name: 'deleted',
              type: 'Boolean',
              nonNull: true,
              defaultValue: false,
              filterable: { default: false },
              generated: true,
              ...(typeof entity.deletable === 'object' && entity.deletable.deleted),
            } satisfies BooleanFieldDefinition,
            {
              name: 'deletedAt',
              type: 'DateTime',
              orderable: true,
              generated: true,
              ...(typeof entity.deletable === 'object' && entity.deletable.deletedAt),
            } satisfies DateTimeFieldDefinition,
            {
              name: 'deletedBy',
              kind: 'relation',
              type: 'User',
              reverse: `deleted${getModelPlural(entity)}`,
              generated: true,
              ...(typeof entity.deletable === 'object' && entity.deletable.deletedBy),
            } satisfies RelationFieldDefinition,
          ]
        : []),
    ];

    for (const field of entity.fields) {
      if (field.kind === 'relation') {
        field.foreignKey = field.foreignKey || `${field.name}Id`;
      }
    }
  }

  for (const model of entities) {
    if (model.parent) {
      const parent = summonByName(entities, model.parent);
      const INHERITED_FIELDS = ['queriable', 'listQueriable', 'creatable', 'updatable', 'deletable'];
      Object.assign(model, pick(parent, INHERITED_FIELDS), pick(model, INHERITED_FIELDS));

      model.fields = [
        ...parent.fields.map((field) => ({
          ...field,
          ...(field.kind === 'relation' &&
            field.reverse && { reverse: field.reverse.replace(getModelPlural(parent), getModelPlural(model)) }),
          ...model.fields.find((childField) => childField.name === field.name),
          inherited: true,
        })),
        ...model.fields.filter((field) => !parent.fields.some((parentField) => parentField.name === field.name)),
      ];
    }
  }

  return definitions;
};

export class Models {
  public models: Model[];
  private modelsByName: Record<string, Model> = {};
  public scalars: ScalarModel[];
  public rawEnums: RawEnumModel[];
  public enums: EnumModel[];
  public inputs: InputModel[];
  public interfaces: InterfaceModel[];
  public objects: ObjectModel[];
  public entities: EntityModel[];
  public definitions: ModelDefinitions;

  constructor(definitions: ModelDefinitions) {
    this.definitions = getFullDefinitions(definitions);

    this.models = this.definitions.map(
      (definition) => new (MODEL_KIND_TO_CLASS_MAPPING[definition.kind] as any)(this, definition),
    );
    for (const model of this.models) {
      this.modelsByName[model.name] = model;
    }
    this.entities = this.models.filter((model): model is EntityModel => model instanceof EntityModel);
    this.enums = this.models.filter((model): model is EnumModel => model instanceof EnumModel);
    this.inputs = this.models.filter((model): model is InputModel => model instanceof InputModel);
    this.interfaces = this.models.filter((model): model is InterfaceModel => model instanceof InterfaceModel);
    this.objects = this.models.filter((model): model is ObjectModel => model instanceof ObjectModel);
    this.rawEnums = this.models.filter((model): model is RawEnumModel => model instanceof RawEnumModel);
    this.scalars = this.models.filter((model): model is ScalarModel => model instanceof ScalarModel);
  }

  public getModel<K extends keyof ModelKindToClassMapping>(name: string, kind?: K): ModelKindToClassMapping[K] {
    const model = get(this.modelsByName, name);
    if (!kind) {
      return model as ModelKindToClassMapping[K];
    }

    const expectedType = MODEL_KIND_TO_CLASS_MAPPING[kind];
    if (!(model instanceof expectedType)) {
      throw new Error(`Model ${name} is not of kind ${kind}.`);
    }

    return model as ModelKindToClassMapping[K];
  }
}

export abstract class Model {
  name: string;
  plural: string;
  description: string;

  constructor(
    public models: Models,
    definition: ModelDefinition,
  ) {
    Object.assign(this, definition);
    this.plural = definition.plural || pluralize(definition.name);
  }
}

export class ScalarModel extends Model {
  kind: 'scalar';
}

export class EnumModel extends Model {
  kind: 'enum';
  values: string[];
  deleted?: true;

  constructor(models: Models, definition: EnumModelDefinition) {
    super(models, definition);
    Object.assign(this, omit(definition, 'name', 'plural', 'description'));
  }
}

export class RawEnumModel extends Model {
  kind: 'raw-enum';
  values: string[];

  constructor(models: Models, definition: RawEnumModelDefinition) {
    super(models, definition);
    Object.assign(this, omit(definition, 'name', 'plural', 'description'));
  }
}

export class InterfaceModel extends Model {
  kind: 'interface';
  fields: EntityField[];

  constructor(models: Models, definition: InterfaceModelDefinition) {
    super(models, definition);
    Object.assign(this, omit(definition, 'name', 'plural', 'description'));
  }
}

export class InputModel extends Model {
  kind: 'model';
  fields: ObjectField[];

  constructor(models: Models, definition: InputModelDefinition) {
    super(models, definition);
    Object.assign(this, omit(definition, 'name', 'plural', 'description'));
  }
}

export class ObjectModel extends Model {
  kind: 'object';
  fields: ObjectField[];

  constructor(models: Models, definition: ObjectModelDefinition) {
    super(models, definition);
    Object.assign(this, omit(definition, 'name', 'plural', 'description'));
  }
}

export class EntityModel extends Model {
  kind: 'entity';
  root?: boolean;
  parent?: string;
  interfaces?: string[];
  queriable?: boolean;
  listQueriable?: boolean;
  creatable?: boolean | { createdBy?: Partial<RelationFieldDefinition>; createdAt?: Partial<DateTimeFieldDefinition> };
  updatable?: boolean | { updatedBy?: Partial<RelationFieldDefinition>; updatedAt?: Partial<DateTimeFieldDefinition> };
  deletable?:
    | boolean
    | {
        deleted?: Partial<BooleanFieldDefinition>;
        deletedBy?: Partial<RelationFieldDefinition>;
        deletedAt?: Partial<DateTimeFieldDefinition>;
      };
  displayField?: string;
  defaultOrderBy?: OrderBy;
  fields: EntityField[];

  // temporary fields for the generation of migrations
  deleted?: true;
  oldName?: string;

  fieldsByName: Record<string, EntityField> = {};
  fieldsByColumnName: Record<string, EntityField> = {};
  private _relations: NormalRelation[];
  private _relationsByName: Record<string, NormalRelation>;
  private _reverseRelations: ReverseRelation[];
  private _reverseRelationsByName: Record<string, ReverseRelation>;
  private _manyToManyRelations: ManyToManyRelation[];
  private _manyToManyRelationsByName: Record<string, ManyToManyRelation>;
  public pluralField: string;
  public slug: string;
  public labelPlural: string;
  public label: string;
  private _parentModel: EntityModel;

  constructor(models: Models, definition: EntityModelDefinition) {
    super(models, definition);
    Object.assign(this, omit(definition, 'name', 'plural', 'description'));
    this.pluralField = typeToField(this.plural);
    this.slug = kebabCase(this.plural);
    this.labelPlural = getLabel(this.plural);
    this.label = getLabel(definition.name);
    for (const field of definition.fields) {
      this.fieldsByName[field.name] = field;
    }
  }

  public getField(name: string) {
    return get(this.fieldsByName, name);
  }

  public get relations() {
    if (!this._relations) {
      this._relations = this.fields
        .filter(isRelation)
        .map((relationField) => new NormalRelation(this, relationField, this.models.getModel(relationField.type, 'entity')));
    }

    return this._relations;
  }

  public get relationsByName() {
    if (!this._relationsByName) {
      this._relationsByName = {};
      for (const relation of this.relations) {
        this._relationsByName[relation.name] = relation;
      }
    }

    return this._relationsByName;
  }

  public getRelation(name: string) {
    return get(this.relationsByName, name);
  }

  public get reverseRelations() {
    if (!this._reverseRelations) {
      this._reverseRelations = this.models.entities.flatMap((model) =>
        model.relations
          .filter((relation) => relation.targetModel.name === this.name || relation.targetModel.name === this.rootModel.name)
          .map((relation) => relation.reverse),
      );
    }

    return this._reverseRelations;
  }

  public get reverseRelationsByName() {
    if (!this._reverseRelationsByName) {
      this._reverseRelationsByName = {};
      for (const reverseRelation of this.reverseRelations) {
        this._reverseRelationsByName[reverseRelation.name] = reverseRelation;
      }
    }

    return this._reverseRelationsByName;
  }

  public getReverseRelation(name: string) {
    return get(this.reverseRelationsByName, name);
  }

  public get manyToManyRelations() {
    if (!this._manyToManyRelations) {
      this._manyToManyRelations = [];
      for (const relationFromSource of this.reverseRelations) {
        const relationToTarget = relationFromSource.targetModel.relations.find(
          (relation) => !relation.field.generated && relation.field.name !== relationFromSource.field.name,
        );
        if (!relationToTarget) {
          continue;
        }

        const inapplicableFields = relationFromSource.targetModel.fields.filter(
          (otherField) =>
            !otherField.generated && ![relationFromSource.field.name, relationToTarget.field.name].includes(otherField.name),
        );
        if (inapplicableFields.length) {
          continue;
        }

        this._manyToManyRelations.push(new ManyToManyRelation(relationFromSource, relationToTarget));
      }
    }

    return this._manyToManyRelations;
  }

  public get manyToManyRelationsByName() {
    if (!this._manyToManyRelationsByName) {
      this._manyToManyRelationsByName = {};
      for (const manyToManyRelation of this.manyToManyRelations) {
        this._manyToManyRelationsByName[manyToManyRelation.name] = manyToManyRelation;
      }
    }

    return this._manyToManyRelationsByName;
  }

  public getManyToManyRelation(name: string) {
    return get(this.manyToManyRelationsByName, name);
  }

  public get parentModel() {
    if (this.parent) {
      if (!this._parentModel) {
        this._parentModel = this.models.getModel(this.parent, 'entity');
      }

      return this._parentModel;
    }
  }

  public get rootModel() {
    return this.parentModel || this;
  }
}

const MODEL_KIND_TO_CLASS_MAPPING = {
  entity: EntityModel,
  enum: EnumModel,
  input: InputModel,
  interface: InterfaceModel,
  object: ObjectModel,
  'raw-enum': RawEnumModel,
  scalar: ScalarModel,
};

type ModelKindToClassMapping = {
  [K in keyof typeof MODEL_KIND_TO_CLASS_MAPPING]: InstanceType<(typeof MODEL_KIND_TO_CLASS_MAPPING)[K]>;
};

export abstract class Relation {
  public searchable: boolean | undefined;
  public label: string;

  constructor(
    public name: string,
    public sourceModel: EntityModel,
    public field: RelationField,
    public targetModel: EntityModel,
  ) {
    this.searchable = targetModel.getField(targetModel.displayField || 'id').searchable;
    this.label = getLabel(name);
  }
}

export class NormalRelation extends Relation {
  public reverse: ReverseRelation;

  constructor(
    sourceModel: EntityModel,
    public field: RelationField,
    targetModel: EntityModel,
  ) {
    super(field.name, sourceModel, field, targetModel);
    this.reverse = new ReverseRelation(this);
  }
}

export class ReverseRelation extends Relation {
  public singularName: string;
  public singularLabel: string;

  constructor(public reverse: NormalRelation) {
    super(
      reverse.field.reverse ||
        (reverse.field.toOne ? typeToField(reverse.sourceModel.name) : reverse.sourceModel.pluralField),
      reverse.targetModel,
      reverse.field,
      reverse.sourceModel,
    );
    this.singularName = singularize(this.name);
    this.singularLabel = getLabel(this.singularName);
  }
}

export class ManyToManyRelation {
  public name: string;
  public label: string;
  public sourceModel: EntityModel;
  public relationFromSource: ReverseRelation;
  public relationModel: EntityModel;
  public relationToTarget: NormalRelation;
  public targetModel: EntityModel;

  constructor(relationFromSource: ReverseRelation, relationToTarget: NormalRelation) {
    this.name = relationFromSource.name;
    this.label = relationFromSource.label;
    this.sourceModel = relationFromSource.sourceModel;
    this.relationFromSource = relationFromSource;
    this.relationModel = relationFromSource.targetModel;
    if (this.relationModel !== relationToTarget.sourceModel) {
      throw new Error(`Relation model is ambiguous.`);
    }
    this.relationToTarget = relationToTarget;
    this.targetModel = relationToTarget.targetModel;
  }
}

const isEntityModelDefinition = (definition: ModelDefinition): definition is EntityModelDefinition =>
  definition.kind === 'entity';

const getModelPlural = (model: EntityModelDefinition) => model.plural || pluralize(model.name);
