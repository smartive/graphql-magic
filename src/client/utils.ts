import upperCase from 'lodash/upperCase';

export const constantCase = (str: string) => upperCase(str).replace(/ /g, '_');
