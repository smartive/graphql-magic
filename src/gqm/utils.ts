import { SourceFile, SyntaxKind, SyntaxList } from 'ts-morph';

export const findDeclarationInFile = (sourceFile: SourceFile, name: string) => {
  const syntaxList = sourceFile.getChildrenOfKind(SyntaxKind.SyntaxList)[0];
  if (!syntaxList) {
    throw new Error('No SyntaxList');
  }
  const declaration = findDeclaration(syntaxList, name);
  if (!declaration) {
    throw new Error('No rawModels declaration');
  }
  return declaration;
};

const findDeclaration = (syntaxList: SyntaxList, name: string) => {
  for (const variableStatement of syntaxList.getChildrenOfKind(SyntaxKind.VariableStatement)) {
    for (const declaration of variableStatement.getDeclarationList().getDeclarations()) {
      if (declaration.getName() === name) {
        return declaration;
      }
    }
  }
};
