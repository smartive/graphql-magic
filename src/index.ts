// created from 'create-ts-index'

export * from './api';
export * from './client';
export * from './db';
export * from './migrations';
export * from './models';
export * from './permissions';
export * from './resolvers';
export * from './schema';
export * from './utils';
export * from './context';
export * from './errors';
export * from './values';
