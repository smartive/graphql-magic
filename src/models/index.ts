// created from 'create-ts-index'

export * from './model-definitions';
export * from './models';
export * from './mutation-hook';
export * from './utils';
