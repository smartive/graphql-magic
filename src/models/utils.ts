import assert from 'assert';
import camelCase from 'lodash/camelCase';
import lodashGet from 'lodash/get';
import startCase from 'lodash/startCase';
import {
  CustomField,
  EntityField,
  EntityModel,
  EnumField,
  EnumModel,
  InputModel,
  InterfaceModel,
  Model,
  ObjectModel,
  PrimitiveField,
  RawEnumModel,
  RelationField,
  ScalarModel,
} from './models';

const isNotFalsy = <T>(v: T | null | undefined | false): v is T => typeof v !== 'undefined' && v !== null && v !== false;

export const merge = <T>(objects: (Record<string, T> | undefined | false)[] | undefined): Record<string, T> =>
  (objects || []).filter(isNotFalsy).reduce((i, acc) => ({ ...acc, ...i }), {});

// Target -> target
export const typeToField = (type: string) => type.substr(0, 1).toLowerCase() + type.substr(1);

export const getLabel = (s: string) => startCase(camelCase(s));

export const or =
  <T>(...predicates: ((field: T) => boolean)[]) =>
  (field: T) =>
    predicates.some((predicate) => predicate(field));

export const and =
  <T>(...predicates: ((field: T) => boolean)[]) =>
  (field: T) =>
    predicates.every((predicate) => predicate(field));

export const not =
  <T>(predicate: (field: T) => boolean) =>
  (field: T) =>
    !predicate(field);

export const isRootModel = (model: EntityModel) => !!model.root;

export const isEntityModel = (model: Model): model is EntityModel => model instanceof EntityModel;

export const isEnumModel = (model: Model): model is EnumModel => model instanceof EnumModel;

export const isRawEnumModel = (model: Model): model is RawEnumModel => model instanceof RawEnumModel;

export const isScalarModel = (model: Model): model is ScalarModel => model instanceof ScalarModel;

export const isObjectModel = (model: Model): model is ObjectModel => model instanceof ObjectModel;

export const isInputModel = (model: Model): model is InputModel => model instanceof InputModel;

export const isInterfaceModel = (model: Model): model is InterfaceModel => model instanceof InterfaceModel;

export const isCreatableModel = (model: EntityModel) => model.creatable && model.fields.some(isCreatableField);

export const isUpdatableModel = (model: EntityModel) => model.updatable && model.fields.some(isUpdatableField);

export const isCreatableField = (field: EntityField) => !field.inherited && !!field.creatable;

export const isUpdatableField = (field: EntityField) => !field.inherited && !!field.updatable;

export const modelNeedsTable = (model: EntityModel) => model.fields.some((field) => !field.inherited);

export const hasName = (name: string) => (field: EntityField) => field.name == name;

export const isPrimitive = (field: EntityField): field is PrimitiveField =>
  field.kind === undefined || field.kind === 'primitive';

export const isEnum = (field: EntityField): field is EnumField => field.kind === 'enum';

export const isRelation = (field: EntityField): field is RelationField => field.kind === 'relation';

export const isInherited = (field: EntityField) => !!field.inherited;

export const isInTable = (field: EntityField) => field.name === 'id' || !field.inherited;

export const isToOneRelation = (field: EntityField): field is RelationField => isRelation(field) && !!field.toOne;

export const isQueriableField = ({ queriable }: EntityField) => queriable !== false;

export const isCustomField = (field: EntityField): field is CustomField => field.kind === 'custom';

export const isVisible = ({ hidden }: EntityField) => hidden !== true;

export const isSimpleField = and(not(isRelation), not(isCustomField));

export const isUpdatable = ({ updatable }: EntityField) => !!updatable;

export const isCreatable = ({ creatable }: EntityField) => !!creatable;

export const isQueriableBy = (role: string) => (field: EntityField) =>
  field.queriable !== false &&
  (field.queriable === undefined ||
    field.queriable === true ||
    !field.queriable.roles ||
    field.queriable.roles.includes(role));

export const isUpdatableBy = (role: string) => (field: EntityField) =>
  field.updatable && (field.updatable === true || !field.updatable.roles || field.updatable.roles.includes(role));

export const isCreatableBy = (role: string) => (field: EntityField) =>
  field.creatable && (field.creatable === true || !field.creatable.roles || field.creatable.roles.includes(role));

export const getActionableRelations = (model: EntityModel, action: 'create' | 'update' | 'filter') =>
  model.relations
    .filter(
      (relation) =>
        relation.field[
          `${action === 'filter' ? action : action.slice(0, -1)}able` as 'filterable' | 'creatable' | 'updatable'
        ],
    )
    .map(({ name }) => name);

export const summonByName = <T extends { name: string }>(array: T[], value: string) => summonByKey(array, 'name', value);

export const summonByKey = <T>(array: readonly T[] | undefined, key: string, value: unknown) =>
  summon(array, (element: T) => lodashGet(element, key) === value, `No element found with ${key} ${value}`);

export const summon = <T>(array: readonly T[] | undefined, cb: Parameters<T[]['find']>[1], errorMessage?: string) => {
  if (array === undefined) {
    console.trace();
    throw new Error('Base array is not defined.');
  }
  const result = array.find(cb);
  if (result === undefined) {
    console.trace();
    throw new Error(errorMessage || 'Element not found.');
  }

  return result;
};

type ForSure<T> = T extends undefined | null ? never : T;

export const it = <T>(object: T | null | undefined): ForSure<T> => {
  if (object === undefined || object === null) {
    console.trace();
    throw new Error('Base object is not defined.');
  }

  return object as ForSure<T>;
};

export const get = <T, U extends keyof ForSure<T>>(object: T | null | undefined, key: U): ForSure<ForSure<T>[U]> => {
  const value = it(object)[key];
  if (value === undefined || value === null) {
    const error = new Error(`Object doesn't have ${String(key)}`);
    console.warn(error);
    throw error;
  }

  return value as ForSure<ForSure<T>[U]>;
};

export const getString = (v: unknown) => {
  assert(typeof v === 'string');

  return v;
};

export const retry = async <T>(cb: () => Promise<T>, condition: (e: any) => boolean) => {
  try {
    return await cb();
  } catch (e) {
    if (condition(e)) {
      return await cb();
    }
    throw e;
  }
};

type Typeof = {
  string: string;
  number: number;
  bigint: bigint;
  boolean: boolean;
  symbol: symbol;
  undefined: undefined;
  object: object;
  function: Function;
};

export const as = <T extends keyof Typeof>(value: unknown, type: T): Typeof[T] => {
  if (typeof value !== type) {
    throw new Error(`No string`);
  }

  return value as Typeof[T];
};
