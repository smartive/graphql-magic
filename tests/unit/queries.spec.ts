import { getEditEntityRelationsQuery } from '../../src';
import { models } from '../utils/models';

describe('queries', () => {
  describe('getEntityRelationsQuery', () => {
    it('applies filters', () => {
      expect(getEditEntityRelationsQuery(models, models.getModel('SomeObject', 'entity'), 'update')).toMatchSnapshot();
    });
  });
});
