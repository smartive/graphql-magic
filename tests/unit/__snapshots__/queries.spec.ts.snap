// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries getEntityRelationsQuery applies filters 1`] = `
"query SelectSomeObjectRelations($anotherWhere: AnotherObjectWhere, $anotherLimit: Int) {
      another: anotherObjects(where: $anotherWhere, limit: $anotherLimit, orderBy: [{ name: ASC }]) {
            id
            display: name
          }
    }"
`;
